### PR TITLE
Fix fusing the dimensions in SliceFlipNormalizePermutePadGpu

### PR DESCRIPTION
**Pull request by staniewzki**
Signed-off-by: Michał Staniewski <m.staniewzki@gmail.com>

## Category:
**Bug fix** (*non-breaking change which fixes an issue*)

## Description:

While working on https://github.com/NVIDIA/DALI/pull/4232, I found a bug in the `SliceFlipNormalizePermutePad` kernel. I added a failing test for an easy reproduction, and a fix that for that. I believe, that the condition I've added is mandatory for the fusing to make any sense, so there should be no regression.

### Affected modules and functionalities:

N/A

### Key points relevant for the review:

The condition in the while loop could probably be optimized/rewritten in a better way.

### Tests:

- [ ] Existing tests apply
- [x] New tests added
  - [ ] Python tests
  - [x] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [ ] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

File: dali/kernels/slice/slice_flip_normalize_permute_pad_gpu.h [hunk 1]
```diff
@@ -223,16 +223,21 @@ class SliceFlipNormalizePermutePadGpu {
...
+      //    - if anchor[last_dim] == 0 && out_shape[last_dim] == in_shape[last_dim],
+      //      then it is not sliced/padded
       // 3. Last dimension is not the channel dimension
-      // 4. The last two dimensions are either both flipped or not flipped
+      // 4. The in_strides of the last dimension and the former one align
+      //    - in_strides[last_dim] * in_shape[last_dim] == in_strides[last_dim - 1]
       int last_dim = Dims - 1;
       while (last_dim > 0 &&
              sample_desc.out_strides[last_dim] ==
                  static_cast<uint64_t>(sample_desc.in_strides[last_dim]) &&
```
Comment (mzient) [inline dali/kernels/slice/slice_flip_normalize_permute_pad_gpu.h:235]: The comment above mentions `abs`, but it's not present in the code. If it's an intentional omission, then the comment should be updated.

**Review by jantonguirao — APPROVED** [review]

**Comment by jantonguirao** [comment]
!build

File: dali/kernels/slice/slice_flip_normalize_permute_pad_gpu.h (hunk 1 above)
Comment (jantonguirao) [reply dali/kernels/slice/slice_flip_normalize_permute_pad_gpu.h:235]: Done

Comment (staniewzki) [reply dali/kernels/slice/slice_flip_normalize_permute_pad_gpu.h:235]: @jantonguirao In my head the abs should make sense there, but when I tried to put that in, test fail, and I was unable to debug it in a limited amount of time.

Let's say that out strides are [5, 1] and in strides are [-5, -1]. We should be able to compress it into out strides [1] and in strides [-1]. The new condition guarantees that the signs are correct.

This isn't a priority in any way, but I'm just raising the issue.

**Review by mzient — APPROVED** [review]

File: dali/kernels/slice/slice_flip_normalize_permute_pad_gpu.h [hunk 2]
```diff
@@ -223,24 +223,36 @@ class SliceFlipNormalizePermutePadGpu {
...
              sample_desc.out_strides[last_dim] ==
-                 static_cast<uint64_t>(sample_desc.in_strides[last_dim]) &&
+                 static_cast<uint64_t>(abs(sample_desc.in_strides[last_dim])) &&
              sample_desc.anchor[last_dim] == 0 &&
              sample_desc.out_shape[last_dim] == sample_desc.in_shape[last_dim] &&
              sample_desc.channel_dim != last_dim &&
-             (sample_desc.in_strides[last_dim] < 0) == (sample_desc.in_strides[last_dim - 1] < 0)) {
+             sample_desc.channel_dim != (last_dim-1) &&
+             sample_desc.in_strides[last_dim] * sample_desc.in_shape[last_dim] ==
+                 sample_desc.in_strides[last_dim - 1]) {
```
Comment (prak-nv) [inline dali/kernels/slice/slice_flip_normalize_permute_pad_gpu.h:245]: This condition is scary, was scary before; Could you please outline it to separate function? As it is, it is incomprehensible for me.

**Review by prak-nv — APPROVED** [review]

**Review by prak-nv — APPROVED** [review]

**Comment by jantonguirao** [comment]
!build

**Pull request merged**